### PR TITLE
[FLINK-20654][FLINK-21104][network] Fix two bugs in the handling of UnalignedCheckpoints

### PR DESCRIPTION
**Pull request by pnowojski**
This PR depends on https://github.com/apache/flink/pull/14797 (@AHeise commits at the bottom)

This PR fixes two bugs. First:
```
    If previous checkpoint is declined, it can happen that task receives both older and newer
    checkpoint barrier on two different channels, before processing any checkpoint cancellation
    message/RPC. If the newer checkpoint barrier happens to be processed before the obsolete one
    incorrect `checkState` in ChannelStatePersister would cause job failure. This checkState
    was assuming that the previous checkpoint would have been aborted/stopped before triggering
    the new one, while in reality, this previous checkpoint has never been triggered on this task
    so it also could not have been stopped.
```
Second:
```
    This commit fixes a bug where RemoteInputChannel was incorrectly deciding which
    buffers should be spilled, if it has received an obsoleted CheckpointBarrier,
    that hasn't been cancelled (yet?).
```
Both commits are tested by the existing UnalignedCheckpointITCase and some freshly added unit tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (**yes** / no / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/ChannelStatePersister.java [hunk 1]
```diff
@@ -67,16 +68,27 @@
         this.channelInfo = checkNotNull(channelInfo);
     }
 
-    protected void startPersisting(long barrierId, List<Buffer> knownBuffers) {
+    protected void startPersisting(long barrierId, List<Buffer> knownBuffers)
+            throws CheckpointException {
         logEvent("startPersisting", barrierId);
-        if (checkpointStatus != CheckpointStatus.BARRIER_RECEIVED && lastSeenBarrier < barrierId) {
+        if (checkpointStatus == CheckpointStatus.BARRIER_RECEIVED && lastSeenBarrier > barrierId) {
+            throw new CheckpointException(
```
Comment (AHeise) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/ChannelStatePersister.java:75]: How does it guarantee that this doesn't cancel a newer checkpoint? I don't see any checkpoint id being bound to the exception.

File: flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/ChannelStatePersisterTest.java [hunk 2]
```diff
@@ -65,38 +83,58 @@ public void testNewBarrierNotOverwrittenByCheckForBarrier() throws IOException {
...
-        // late barrier
-        persister.checkForBarrier(barrier(1L));
+    @Test
+    public void testLateBarrierOnNotYetCancelledCheckpoint() throws Exception {
+        testLateBarrier(false, false);
+    }
 
-        persister.startPersisting(2L, Collections.emptyList());
-        persister.checkForBarrier(barrier(2L));
+    public void testLateBarrier(
```
Comment (AHeise) [inline flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/ChannelStatePersisterTest.java:100]: nit: private

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java [hunk 3]
```diff
@@ -638,6 +637,11 @@ private void notifyPriorityEventForce() {
                     CheckpointFailureReason
                             .CHECKPOINT_SUBSUMED); // currently, at most one active unaligned
             // checkpoint is possible
+        } else if (checkpointId > lastBarrierId) {
+            // This channel has received some obsolete barrier, older compared to the checkpointId
+            // which we are processing right now, and we should ignore that obsoleted checkpoint
+            // barrier sequence number.
+            resetLastBarrier();
```
Comment (AHeise) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java:644]: I wouldn't have had a side-effect here and just returned an empty list (which I'd also return with side-effect).

**Review by AHeise — COMMENTED** [review]
In general good catches and great tests. I still have two issues that I'd like to see addressed.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/ChannelStatePersister.java (hunk 1 above)
Comment (pnowojski) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/ChannelStatePersister.java:75]: As discussed off-line, it's cancelling the currently triggered exception

**Comment by AHeise** [comment]
Incorporated into #14797 .

**Pull request closed without merging**